File: backend/utils/variable_resolver.py
"""
Enhanced variable resolution system for workflow execution
"""
import re
import logging
from typing import Dict, Any, List, Tuple, Optional

logger = logging.getLogger("workflow_api")

class VariableResolver:
    """Handles variable resolution in workflow execution"""
    
    def __init__(self, nodes: List[Dict], edges: List[Dict]):
        self.nodes = nodes
        self.edges = edges
        self.node_name_map = self._create_node_name_map()
        
    def _create_node_name_map(self) -> Dict[str, str]:
        """Create bidirectional mapping between node IDs and display names"""
        node_name_map = {}
        
        for node in self.nodes:
            node_id = node["id"]
            node_params = node.get("data", {}).get("params", {})
            
            # Get node name from params or generate default
            node_name = node_params.get("nodeName")
            if not node_name:
                node_type = node["type"]
                if node_id.startswith(f"{node_type}_"):
                    node_name = node_id
                else:
                    # Extract index from ID
                    parts = node_id.split('-')
                    index = parts[-1] if len(parts) > 1 else '0'
                    node_name = f"{node_type}_{index}"
            
            # Bidirectional mapping
            node_name_map[node_id] = node_name
            node_name_map[node_name] = node_id
            
        return node_name_map
    
    def standardize_node_output(self, node_type: str, output: Any) -> Dict[str, Any]:
        """Standardize node output fields based on node type"""
        if not isinstance(output, dict):
            return {"output": output}
        
        standardized = output.copy()
        
        # Ensure all nodes have an 'output' field
        if "output" not in standardized:
            if "response" in standardized:
                standardized["output"] = standardized["response"]
            elif "text" in standardized:
                standardized["output"] = standardized["text"]
            elif "content" in standardized:
                standardized["output"] = standardized["content"]
            elif "result" in standardized:
                standardized["output"] = standardized["result"]
            else:
                # Find first string value
                for key, value in standardized.items():
                    if isinstance(value, str) and value:
                        standardized["output"] = value
                        break
                else:
                    standardized["output"] = ""
        
        # AI nodes should have 'response' field
        ai_node_types = [
            "openai", "anthropic", "claude35", "gemini", "cohere", 
            "perplexity", "xai", "aws", "azure"
        ]
        if node_type in ai_node_types and "response" not in standardized:
            standardized["response"] = standardized.get("output", "")
        
        # Input nodes should have 'text' field for compatibility
        if node_type == "input" and "text" not in standardized:
            standardized["text"] = standardized.get("output", "")
            
        return standardized
    
    def resolve_variables(self, text: str, node_outputs: Dict[str, Any]) -> str:
        """Resolve variables in text using {{ nodeName.field }} format"""
        if not isinstance(text, str):
            return text
        
        # Pattern to match {{ nodeName.field }} with optional whitespace
        pattern = r'\{\{\s*([^}]+)\s*\}\}'
        
        def replace_variable(match):
            variable = match.group(1).strip()
            
            # Split by dot to get node name and field
            if '.' in variable:
                node_name, field = variable.split('.', 1)
                node_name = node_name.strip()
                field = field.strip()
            else:
                node_name = variable.strip()
                field = 'output'
            
            # Find node output
            node_output = self._find_node_output(node_name, node_outputs)
            
            if node_output and isinstance(node_output, dict):
                # Get requested field
                if field in node_output:
                    value = node_output[field]
                    return str(value) if value is not None else ""
                else:
                    # Try field alternatives
                    alternatives = self._get_field_alternatives(field)
                    for alt in alternatives:
                        if alt in node_output:
                            value = node_output[alt]
                            return str(value) if value is not None else ""
            
            # Log warning and return original if not resolved
            logger.warning(f"Could not resolve variable: {variable}")
            return match.group(0)
        
        return re.sub(pattern, replace_variable, text)
    
    def _find_node_output(self, node_name: str, node_outputs: Dict[str, Any]) -> Optional[Dict[str, Any]]:
        """Find node output by name or ID"""
        # Direct lookup
        if node_name in node_outputs:
            return node_outputs[node_name]
        
        # Try ID lookup
        node_id = self.node_name_map.get(node_name)
        if node_id and node_id in node_outputs:
            return node_outputs[node_id]
        
        # Reverse lookup
        if node_name in self.node_name_map:
            mapped_name = self.node_name_map[node_name]
            if mapped_name in node_outputs:
                return node_outputs[mapped_name]
        
        return None
    
    def _get_field_alternatives(self, field: str) -> List[str]:
        """Get alternative field names for common mappings"""
        alternatives = {
            'text': ['output', 'response', 'content'],
            'response': ['output', 'text', 'content'],
            'output': ['response', 'text', 'content', 'result'],
            'content': ['output', 'response', 'text'],
            'result': ['output', 'response', 'content']
        }
        return alternatives.get(field, [])
    
    def validate_variables(self, text: str, available_outputs: Dict[str, Any]) -> List[Dict[str, Any]]:
        """Validate that variables in text can be resolved"""
        if not isinstance(text, str):
            return []
        
        pattern = r'\{\{\s*([^}]+)\s*\}\}'
        matches = re.findall(pattern, text)
        validation_results = []
        
        for variable in matches:
            variable = variable.strip()
            node_name, field = (variable.split('.', 1) if '.' in variable else (variable, 'output'))
            node_name = node_name.strip()
            field = field.strip()
            
            # Check if variable can be resolved
            node_output = self._find_node_output(node_name, available_outputs)
            is_valid = node_output and (
                field in node_output or 
                any(alt in node_output for alt in self._get_field_alternatives(field))
            )
            
            suggestion = None
            if not is_valid and node_output:
                available_fields = list(node_output.keys())
                if available_fields:
                    suggestion = f"{node_name}.{available_fields[0]}"
            
            validation_results.append({
                "variable": variable,
                "is_valid": is_valid,
                "suggestion": suggestion
            })
        
        return validation_results

File: src/utils/variableResolver.ts
/**
 * Enhanced variable resolution utilities for frontend
 */

export interface NodeOutput {
  [key: string]: any;
}

export interface VariableContext {
  nodeOutputs: Record<string, NodeOutput>;
  nodeNameMap: Record<string, string>;
}

export interface VariableValidationResult {
  variable: string;
  isValid: boolean;
  suggestion?: string;
  nodeExists?: boolean;
  fieldExists?: boolean;
}

/**
 * Resolve variables in text using the format {{ nodeName.field }}
 */
export function resolveVariables(
  text: string, 
  context: VariableContext
): string {
  if (!text || typeof text !== 'string') {
    return text;
  }

  // Pattern to match {{ nodeName.field }} with optional whitespace
  const pattern = /\{\{\s*([^}]+)\s*\}\}/g;
  
  return text.replace(pattern, (match, variable) => {
    const trimmedVariable = variable.trim();
    
    // Split by dot to get node name and field
    let nodeName: string;
    let field: string;
    
    if (trimmedVariable.includes('.')) {
      [nodeName, field] = trimmedVariable.split('.', 2);
      nodeName = nodeName.trim();
      field = field.trim();
    } else {
      // If no field specified, assume 'output'
      nodeName = trimmedVariable;
      field = 'output';
    }
    
    // Try to find the node output
    const nodeOutput = findNodeOutput(nodeName, context);
    
    if (nodeOutput && typeof nodeOutput === 'object') {
      // Get the requested field
      if (field in nodeOutput) {
        const value = nodeOutput[field];
        return value !== null && value !== undefined ? String(value) : '';
      } else {
        // Try common field alternatives
        const alternatives = getFieldAlternatives(field);
        for (const alt of alternatives) {
          if (alt in nodeOutput) {
            const value = nodeOutput[alt];
            return value !== null && value !== undefined ? String(value) : '';
          }
        }
      }
    }
    
    // If variable couldn't be resolved, return the original placeholder
    console.warn(`Could not resolve variable: ${trimmedVariable}`);
    return match; // Return original {{ variable }}
  });
}

/**
 * Find node output by name or ID
 */
function findNodeOutput(
  nodeName: string, 
  context: VariableContext
): NodeOutput | null {
  const { nodeOutputs, nodeNameMap } = context;
  
  // First try direct node name lookup
  if (nodeName in nodeOutputs) {
    return nodeOutputs[nodeName];
  }
  
  // Try to find by node ID if node name lookup fails
  const nodeId = nodeNameMap[nodeName];
  if (nodeId && nodeId in nodeOutputs) {
    return nodeOutputs[nodeId];
  }
  
  // Try reverse lookup (in case nodeName is actually an ID)
  if (nodeName in nodeNameMap) {
    const mappedName = nodeNameMap[nodeName];
    if (mappedName in nodeOutputs) {
      return nodeOutputs[mappedName];
    }
  }
  
  return null;
}

/**
 * Get alternative field names for common field mappings
 */
function getFieldAlternatives(field: string): string[] {
  const alternatives: Record<string, string[]> = {
    'text': ['output', 'response', 'content'],
    'response': ['output', 'text', 'content'],
    'output': ['response', 'text', 'content', 'result'],
    'content': ['output', 'response', 'text'],
    'result': ['output', 'response', 'content']
  };
  
  return alternatives[field] || [];
}

/**
 * Standardize node output fields based on node type
 */
export function standardizeNodeOutput(nodeType: string, output: any): NodeOutput {
  if (!output || typeof output !== 'object') {
    return { output: output || '' };
  }
  
  const standardized = { ...output };
  
  // Ensure all nodes have an 'output' field
  if (!('output' in standardized)) {
    // Map common field names to 'output'
    if ('response' in standardized) {
      standardized.output = standardized.response;
    } else if ('text' in standardized) {
      standardized.output = standardized.text;
    } else if ('content' in standardized) {
      standardized.output = standardized.content;
    } else if ('result' in standardized) {
      standardized.output = standardized.result;
    } else {
      // Find the first string value
      for (const [key, value] of Object.entries(standardized)) {
        if (typeof value === 'string' && value) {
          standardized.output = value;
          break;
        }
      }
      
      // If still no output, set empty string
      if (!('output' in standardized)) {
        standardized.output = '';
      }
    }
  }
  
  // For AI nodes, ensure they have a 'response' field
  const aiNodeTypes = [
    'openai', 'anthropic', 'claude35', 'gemini', 'cohere', 
    'perplexity', 'xai', 'aws', 'azure'
  ];
  
  if (aiNodeTypes.includes(nodeType)) {
    if (!('response' in standardized)) {
      standardized.response = standardized.output;
    }
  }
  
  // For input nodes, ensure they have a 'text' field for backward compatibility
  if (nodeType === 'input') {
    if (!('text' in standardized)) {
      standardized.text = standardized.output;
    }
  }
  
  return standardized;
}

/**
 * Create a mapping between node IDs and their display names
 */
export function createNodeNameMap(nodes: any[]): Record<string, string> {
  const nodeNameMap: Record<string, string> = {};
  
  for (const node of nodes) {
    const nodeId = node.id;
    const nodeParams = node.data?.params || {};
    
    // Get the node name from params, or generate a default
    let nodeName = nodeParams.nodeName;
    if (!nodeName) {
      // Generate default name based on type and ID
      const nodeType = node.type;
      if (nodeId.startsWith(`${nodeType}_`)) {
        nodeName = nodeId;
      } else {
        // Extract index from ID if possible
        const parts = nodeId.split('-');
        const index = parts.length > 1 ? parts[parts.length - 1] : '0';
        nodeName = `${nodeType}_${index}`;
      }
    }
    
    // Map both directions for lookups
    nodeNameMap[nodeId] = nodeName;
    nodeNameMap[nodeName] = nodeId;
  }
  
  return nodeNameMap;
}

/**
 * Validate that all variables in text can be resolved
 */
export function validateVariables(
  text: string, 
  context: VariableContext
): VariableValidationResult[] {
  if (!text || typeof text !== 'string') {
    return [];
  }

  const pattern = /\{\{\s*([^}]+)\s*\}\}/g;
  const matches = [...text.matchAll(pattern)];
  const validationResults: VariableValidationResult[] = [];
  
  for (const match of matches) {
    const variable = match[1].trim();
    const [nodeName, field = 'output'] = variable.includes('.') 
      ? variable.split('.', 2) 
      : [variable, 'output'];
    
    // Check if node exists
    const nodeOutput = findNodeOutput(nodeName.trim(), context);
    const nodeExists = nodeOutput !== null;
    
    // Check if field exists
    let fieldExists = false;
    if (nodeOutput && typeof nodeOutput === 'object') {
      fieldExists = field.trim() in nodeOutput || 
        getFieldAlternatives(field.trim()).some(alt => alt in nodeOutput);
    }
    
    const isValid = nodeExists && fieldExists;
    
    let suggestion = undefined;
    if (!isValid && nodeOutput) {
      // Suggest available fields
      const availableFields = Object.keys(nodeOutput);
      if (availableFields.length > 0) {
        suggestion = `${nodeName}.${availableFields[0]}`;
      }
    }
    
    validationResults.push({
      variable,
      isValid,
      suggestion,
      nodeExists,
      fieldExists
    });
  }
  
  return validationResults;
}

/**
 * Get available variables from nodes
 */
export function getAvailableVariables(nodes: any[]): Array<{
  nodeId: string;
  nodeName: string;
  nodeType: string;
  fields: Array<{name: string, type: string, description: string}>;
}> {
  return nodes.map(node => {
    const nodeId = node.id;
    const nodeName = node.data?.params?.nodeName || nodeId;
    const nodeType = node.type;
    
    // Define fields based on node type
    let fields: Array<{name: string, type: string, description: string}> = [];
    
    switch (nodeType) {
      case 'input':
        fields = [
          { name: 'output', type: 'Text', description: 'User input text' },
          { name: 'text', type: 'Text', description: 'User input text (legacy)' }
        ];
        break;
      
      case 'openai':
      case 'anthropic':
      case 'claude35':
      case 'gemini':
      case 'cohere':
      case 'perplexity':
      case 'xai':
      case 'aws':
      case 'azure':
        fields = [
          { name: 'response', type: 'Text', description: 'AI model response' },
          { name: 'output', type: 'Text', description: 'AI model output' },
          { name: 'model', type: 'Text', description: 'Model name used' },
          { name: 'input_tokens', type: 'Number', description: 'Input tokens consumed' },
          { name: 'output_tokens', type: 'Number', description: 'Output tokens generated' }
        ];
        break;
      
      case 'transform':
      case 'scripts':
        fields = [
          { name: 'output', type: 'Text', description: 'Transformed text' },
          { name: 'transformed_text', type: 'Text', description: 'Transformed text (legacy)' }
        ];
        break;
      
      case 'document-to-text':
        fields = [
          { name: 'output', type: 'Text', description: 'Extracted text from document' }
        ];
        break;
      
      case 'kb-search':
        fields = [
          { name: 'results', type: 'Array', description: 'Search results' },
          { name: 'output', type: 'Text', description: 'Formatted results' },
          { name: 'metadata', type: 'Object', description: 'Search metadata' }
        ];
        break;
      
      default:
        fields = [
          { name: 'output', type: 'Text', description: 'Node output' }
        ];
    }
    
    return {
      nodeId,
      nodeName,
      nodeType,
      fields
    };
  });
}

export default {
  resolveVariables,
  standardizeNodeOutput,
  createNodeNameMap,
  validateVariables,
  getAvailableVariables
};